udata/harvest/backends/dcat.py

```diff
@@ -146,25 +157,51 @@ def walk_graph(self, url: str, fmt: str) -> Generator[tuple[int, Graph], None, N
                     break
 
             yield page_number, subgraph
-            if self.is_done():
+            if self.has_reached_max_items():
                 return
 
             page_number += 1
 
     def process_one_datasets_page(self, page_number: int, page: Graph):
         for node in page.subjects(RDF.type, DCAT.Dataset):
             remote_id = page.value(node, DCT.identifier)
+            if self.is_dataset_external_to_this_page(page, node):
+                continue
+
             self.process_dataset(remote_id, page_number=page_number, page=page, node=node)
 
-            if self.is_done():
+            if self.has_reached_max_items():
                 return
 
+    def is_dataset_external_to_this_page(self, page: Graph, node) -> bool:
+        # In dataservice nodes we have `servesDataset` or `hasPart` that can contains nodes
+        # with type=dataset. We don't want to process them because these nodes are empty (they
+        # only contains a link to the dataset definition).
+        # These datasets are either present in the catalog in previous or next pages or
+        # external from the catalog we are currently harvesting (so we don't want to harvest them).
+        # First we thought of skipping them inside `dataset_from_rdf` (see :ExcludeExternalyDefinedDataset)
+        # but it creates a lot of "fake" items in the job and raising problems (reaching the max harvest item for
+        # example and not getting to the "real" datasets/dataservices in subsequent pages)
+        # So to prevent creating a lot of useless items in the job we check here to see if there is no title and
+        # if `isPrimaryTopicOf` is present.
+        # `isPrimaryTopicOf` is the tag present in the first harvester raising the problem, it may exists other
+        # values of the same sort we need to check here.
+        # This is not dangerous because we check for missing title in `dataset_from_rdf` later so we would have skipped
+        # this dataset anyway.
+
+        resource = page.resource(node)
+
+        title = rdf_value(resource, DCT.title)
+        primary_topic = rdf_value(resource, FOAF.isPrimaryTopicOf)
```

Comment:
I don't trust the `isPrimaryTopicOf` heuristic approach. I think all datasets described by GeoNetwork and transformed with the XSLT end up with this property. It's not only the case in the `servesDataset` context.
See for example this [random dataset DCAT output](https://geodcat-ap.semic.eu/api/?outputSchema=core&src=https%3A%2F%2Fcatalogue-formation.sigloire.fr%2Fgeonetwork%2Fsrv%2Fapi%2Frecords%2Ffr-120066022-jdd-73fbab7e-e7cf-4b12-b88b-5b91efc7b6de%2Fformatters%2Fxml&outputFormat=text%2Fhtml).

Ideally, we would like to exclude datasets that are *only* predicate of `servesDataset`  and `hasPart` explicitly?
Indeed, they are not directly bound to the harvested catalogue at this moment, we wouldn't want to harvest them if they only appear as reference only?

Reply:
I didn't find another way to filter these nodes. But I don't exclude all the nodes with `isPrimaryTopicOf`, only nodes without title (so they will be skipped anyway), I don't think there is a lot of datasets without titles and `isPrimaryTopicOf` that are real datasets?

Reply:
I prefer the alternative done in https://github.com/opendatateam/udata/pull/3285/commits/ca1bc595dcafbd9e818405336fe29ccd86752a05, in conjunction with a check on a missing title